src/software/ai/rl/dqn.hpp

```diff
@@ -0,0 +1,212 @@
+}
+
+template <typename TState, typename TAction>
+torch::Tensor DQN<TState, TAction>::update(
+    const std::vector<Transition<TState, TAction>>& batch, torch::Tensor weights)
+{
+    const int batch_size           = static_cast<int>(batch.size());
+    const int state_size           = static_cast<int>(TState::size());
+    torch::Tensor state_batch      = torch::empty({batch_size, state_size});
+    torch::Tensor action_batch     = torch::empty({batch_size}, {torch::kInt64});
+    torch::Tensor reward_batch     = torch::empty({batch_size});
+    torch::Tensor next_state_batch = torch::empty({batch_size, state_size});
+    torch::Tensor done_batch       = torch::empty({batch_size}, {torch::kBool});
+
+    for (size_t i = 0; i < batch.size(); ++i)
+    {
+        const Transition<TState, TAction>& transition = batch.at(i);
+
+        state_batch[i]      = transition.state.getTensor();
+        action_batch[i]     = static_cast<int>(transition.action);
+        reward_batch[i]     = transition.reward;
+        next_state_batch[i] = transition.next_state.getTensor();
+        done_batch[i]       = transition.done;
+    }
+
+    // Compute Q(s, a) using the current network
+    torch::Tensor q_values =
+        current_net_->forward(state_batch).gather(1, action_batch.unsqueeze(1)).squeeze();
+
+    // Compute argmax_{a}(Q(s', a)) using the target network
+    torch::Tensor next_q_values;
+    {
+        torch::NoGradGuard no_grad;
+        next_q_values = std::get<0>(target_net_->forward(next_state_batch).max(1));
+    }
+
+    // Set next_q_values to 0 where s' is the final state
+    // (no more states after s', so Q(s', a) should only give the immediate reward)
+    next_q_values.index_put_({done_batch}, 0);
+
+    // Compute target value of Q(s, a)
+    torch::Tensor target_q_values = reward_batch + (next_q_values * discount_rate_);
+
+    // Compute TD error
+    torch::Tensor td_error = torch::abs(target_q_values - q_values).detach();
+
+    // Compute Huber loss
+    torch::Tensor loss =
+        torch::nn::functional::huber_loss(q_values * weights, target_q_values * weights);
+
+    // Optimize current network
+    optimizer_.zero_grad();
+    loss.backward();
+    torch::nn::utils::clip_grad_value_(current_net_->parameters(), CLIP_GRADIENT_VALUE);
+    optimizer_.step();
+
+    // Soft update of the target network's weights
+    // θ' = τ * θ + (1 + τ) * θ'
+    {
+        torch::NoGradGuard no_grad;
+        const auto& current_params = current_net_->named_parameters();
+        for (const auto& pair : target_net_->named_parameters())
+        {
+            const torch::Tensor* current_param = current_params.find(pair.key());
+            if (current_param != nullptr)
+            {
+                const torch::Tensor& target_param = pair.value();
+                target_param.copy_(soft_update_tau_ * current_param->clone() +
+                                   (1 - soft_update_tau_) * target_param);
```

Comment:
just checking... `- soft_update_tau_` or `+ soft_update_tau_`. Didn't seem to match the equation in the comment.

Reply:
Good catch, 1 - tau is correct